perf_bench_test.go

```diff
@@ -24,127 +24,619 @@ package main
-			vmGroup.Add(1)
+	// the constants for metric names
+	const (
+		avgExecTime = "average-execution-time"
+		rpsPerCore  = "RPS-per-Core"
+	)
 
-			funcName := funcs[vmID%len(funcs)]
-			imageName := images[funcName]
-			vmIDString := strconv.Itoa(vmID)
+	for step := stepSize; step < 1+stepSize; step += stepSize {
+		var (
+			vmID, requestID             int
+			invokExecTime, realRequests int64
+			serveMetric                 = metrics.NewMetric()
+			rps                         = int64(step * float64(targetRPS))
+			totalRequests               = injectDuration * float64(rps)
+			remainingRequests           = totalRequests
+			profiler                    = profile.NewProfiler(injectDuration, *profilerInterval, vmNum, *profilerLevel,
+				*profilerMetrics, "profile", *isBindSocket)
+		)
 
-			tStart := time.Now()
-			go serveVM(t, tStart, vmIDString, imageName, &vmGroup, isSyncOffload)
+		if rps <= 0 {
+			continue
+		}
+
+		ticker := time.NewTicker(time.Duration(time.Second.Nanoseconds() / rps))
+
+		log.Infof("Current RPS: %d", rps)
+
+		tStart := time.Now()
+		latencyCh := make(chan LatencyStat)
+		go measureTailLatency(t, vmNum, images, &isProfile, latencyCh)
+
+		err := profiler.Run()
+		require.NoError(t, err, "Run profiler returned error")
+		go configureProfiler(&isProfile, profiler)
 
-			vmID = (vmID + 1) % *vmNum
-		case <-done:
-			ticker.Stop()
+		for remainingRequests > 0 {
+			if tickerT := <-ticker.C; !tickerT.IsZero() {
+				vmGroup.Add(1)
+				remainingRequests--
+
+				imageName := images[vmID%len(images)]
+
+				go loadVMs(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
+				requestID++
+
+				vmID = (vmID + 1) % vmNum
+			}
 		}
+		ticker.Stop()
+		vmGroup.Wait()
+		log.Debugf("All VM returned in %d Milliseconds", time.Since(tStart).Milliseconds())
+		latencies := <-latencyCh
+		log.Debugf("Mean Latency: %f, Tail Latency: %f", latencies.meanLatency, latencies.tailLatency)
+
+		// Collect results
+		serveMetric.MetricMap[avgExecTime] = float64(invokExecTime) / float64(realRequests)
+		serveMetric.MetricMap[rpsPerCore] = float64(realRequests) / (profiler.GetCoolDownTime() - profiler.GetWarmUpTime())
+		if cores > vmNum {
+			serveMetric.MetricMap[rpsPerCore] /= float64(vmNum)
+		} else {
+			serveMetric.MetricMap[rpsPerCore] /= float64(cores)
+		}
+		result, err := profiler.GetResult()
+		profiledCores := profiler.GetCores()
+		log.Debugf("%d cores are recorded: %v", len(profiledCores), profiledCores)
+		require.NoError(t, err, "Stopping profiler returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f", eventName, value)
+			serveMetric.MetricMap[eventName] = value
+		}
+		log.Debugf("%s: %f", avgExecTime, serveMetric.MetricMap[avgExecTime])
+		log.Debugf("%s: %f", rpsPerCore, serveMetric.MetricMap[rpsPerCore])
+		profiler.PrintBottlenecks()
+
+		// if tail latency violates the contraints that it should be less than 5x service time,
```

Comment:
it's not 5x, it's according to the threshold

Reply:
not constraint but threshold

Reply:
Done